### DISCUSSION TITLE
How to get new values each test

### DISCUSSION THREAD

**Discussion by AndreLeBlanc** [dubzzz/fast-check#4972] [Q&A]
I am writing tests where data is put into a database with unique constraints. customerEmail has to be a unique value for example. Is there a way to write a test so that if I do 100 runs I won't get the same value?

Some JSONs I am generating have up to 10 unique constraints, and with a hundred test-runs I am regularly getting test failures.

**Comment by gruhn** [comment]
Sorry, if I'm misunderstanding but does your test look somewhat like this?
```typescript
fc.assert(
  fc.property(
    fc.emailAddress(),
    customerEmail => {
      // ...
      database.insert({ customerEmail })
      // ...
    }
  )
)
```
Because the strong assumption with these tests is that **no state persists across test runs**. Otherwise this can badly mess with shrinking. For example, let's say the test fails (for whatever reason) when there is a customer email in the database that contains a "+". One test run inserts `dfjadlksfjdkf+ajdfalsdfjkaljfa@examfajdasfkk.fjk` provoking the issue. Now the shrinker tries to find a simpler counterexample but it can suddenly reproduce the issue for any input email address, so in the end you get the counterexample `a@a.aa` which is too simple and doesn't actually reproduce the issue in isolation. 

The test above checks "what happens if I insert a single record into the database" in a many different ways. But it sounds like what you really want to test is "what happens if I insert many records into the database". For that I would write the test like this:
```typescript
fc.assert(
  fc.property(
    fc.uniqueArray(fc.emailAddress()),
    customerEmails => {
      database.reset() 
      // ...
      for (const customerEmail of customerEmails) {
        database.insert({ customerEmail })
      }
      // ...
    }
  )
)
```
Now I can also use `fc.uniqueArray` to make sure I get pairwise different email addresses. Just make sure to reset the database as good as possible after or before each test run, so no state is carried over.

**Comment by dubzzz** [comment]
We provide a flag called `ignoreEqualValues` that could be passed to `fc.assert` to avoid receiving equal values.

Usually I would rather advocate for properties being able to support duplicates. When they don't support such a constraint they defeat one of the key pillars of property based testing: properties being pure.

That said maybe this flag will work for you.
